// jsg/annotate.js
var idents = require('javascript-idents');
var infer = require('tern/lib/infer');
var nodejs_util = require('./nodejs_util');
var path = require('path');
var tern = require('tern');
var symbol_id = require('./symbol_id');

// Returns all references detected by tern in the JavaScript files (origins).
exports.refs = function(origins, options) {
  if (typeof origins == 'string') origins = [origins];
  var state = new State(origins, options || {});

  var numIdents = 0, numResolved = 0;
  state.cx.parent.files.forEach(function(file) {
    if (!state.isTarget(file.name)) return;
    idents.inspect(file.ast, function(ident) {
      numIdents++;
      var resolved = resolve(file, ident, state);
      if (resolved) numResolved++;
    });
  });

  if (options.stats) {
    console.error('Resolved ' + numResolved + '/' + numIdents + ' idents (' + (100*numResolved/numIdents).toFixed(1) + '%)');
  }

  return state.output;
};

function State(origins, options) {
  this.origins = origins;
  this.cx = infer.cx();
  this.output = [];
}

State.prototype.isTarget = function(origin) {
  return this.origins.indexOf(origin) > -1;
};

function resolve(file, ident, state) {
  var out = {file: file.name, span: ident.start + '-' + ident.end};

  out.target = getRefTarget(file, ident);
  if (out.target) {
    if (state.isTarget(out.target.origin)) delete out.target.origin;
  }

  out.def = ident._isDef;

  if (out.target) {
    state.output.push(out);
    return true;
  }
  return false;
}

function getRefTarget(file, ident) {
  var expr;
  try {
    expr = tern.findQueryExpr(file, ident);
  } catch (e) {
    console.error('No expression at ' + file.name + ':' + ident.start + '-' + ident.end);
    return null;
  }

  return getTarget(infer.expressionType(expr));
}

var getTarget = exports.getTarget = function (av) {
  if (av && av.path && av.path.match(/^Module\.prototype\.parent/)) {
    return {
      abstract: false,
      module: 'module',
      namespace: 'commonjs',
      path: av.path.replace(/^Module/, "module")
    };
  }

  if (av._path) return getConcretePathTypeID(av._path);
  var type = av.getType(false);
  if (!type) return null;
  return getTypeID(type);
}

function getConcretePathTypeID(symbolPath) {
  var target = symbol_id.parse(symbolPath);
  target.abstract = false;

  var server = infer.cx().parent;
  if (server._node && server._node.options.coreModulesDir) {
    var isInNodeCoreModule = path.resolve(path.dirname(target.module)) == path.resolve(server._node.options.coreModulesDir);
    if (isInNodeCoreModule) target.nodejsCoreModule = path.basename(target.module, '.js');
  }

  if (target.namespace == 'commonjs') {
    nodejs_util.addPackageInfo(target);
  }

  return target;
}

function getTypeID(type) {
  // Hack for CommonJS "module"
  if (type.name == 'Module' && type.proto.origin == 'node' && type.proto.name == 'Module.prototype') {
    type.origin = 'node';
    type._isCommonJSModule = true;
  }

  var target = {origin: type.origin};
  switch (type.origin) {
  case 'ecma5':
  case 'browser':
    target.abstract = true;
    // Using the type name is usually what we want, but it may over-resolve.
    // Uncomment this line below to see where it's resolving.
    //
    // if (type.path != type.name) console.error('type.path != type.name:   ', type.path, '!=', type.name);
    target.path = type.name;
    target.namespace = 'global';
    break;
  case 'node':
    target.abstract = true;

    // Hack for CommonJS "module"
    if (type._isCommonJSModule) {
      target.module = 'module';
      target.namespace = 'global';
      break;
    }

    // Hack for CommonJS "require"
    if (type.name == 'require') {
      type.path = type.name = 'module.prototype.require';
    } else if (type.name.indexOf('require.') == 0) {
      type.name = type.name.replace(/^require\./, 'module.prototype.require.');
    }

    if (/^console.[a-z]+/.test(type.name)) {
      type.name = type.name.replace(/^console\./, 'console.Console.prototype.');
    }
    var parts = type.name.split('.');
    target.namespace = 'commonjs';
    target.module = parts[0];
    target.path = parts.slice(1).join('.');
    break;
  case 'requirejs':
    target.abstract = true;
    target.path = type.path;
    target.namespace = 'global';
    target.module = '';
    break;
  default:
    target.abstract = false;

    // Hack for CommonJS module obtained via "require"
    if (type.metaData && type.metaData.nodejs && type.metaData.nodejs.moduleExports) {
      type.path = '!commonjs.' + type.origin.replace(/\./g, '`');
    }

    // Hack for RequireJS module definition
    if (type.metaData && type.metaData.amd && type.metaData.amd.module) {
      type.path = '!requirejs.' + type.origin.replace(/\./g, '`');
    }

    if (!type.path) {
      // Type is probably a local variable or some other type that's unreachable
      // from the topScope or roots.
      // throw new Error('no type.path: ' + require('util').inspect(type));
      return;
    }
    return getConcretePathTypeID(type.path);
  }

  return target;
}
