// jsg/dump.js
var defnode = require('defnode');
var scopewalk = require('tern-scope-walk');
var infer = require('tern/lib/infer');
var symbol_id = require('./symbol_id');

// Returns all def paths in the specified files (origins) by walking the tern
// scope hieriarchy. After walking, also runs functions in
// passes.postDumpScopeWalk (passing state as an argument), which tern plugins
// can register for (see graph.js).
exports.dump = function(origins, options) {
  var state = new State(origins, options);
  var allOrigins = state.cx.parent.files.map(function(f) { return f.name; }).sort();

  var avals = scopewalk.walk(allOrigins);
  for (var path in avals) {
    var aval = avals[path];
    propagateAndStoreExplicitly(aval);
    if (!state.isTarget(aval.origin)) delete avals[path];
  }

  runPass(state.passes.postDumpScopeWalk, state);

  var sortedPaths = Object.keys(avals).sort();

  sortedPaths.forEach(function(path) {
    state.output.push(describe(path, avals[path], state));
  });

  return state.output;
};

function State(origins, options) {
  this.origins = origins;
  this.cx = infer.cx();
  this.passes = options.passes || this.cx.parent && this.cx.parent.passes || {};
  this.output = [];
  this.options = options;
}

State.prototype.isTarget = function(origin) {
  return this.origins.indexOf(origin) > -1;
}

State.prototype.getSpan = function(node) {
  return node.start + '-' + node.end;
};


function hop(obj, prop) {
  return Object.prototype.hasOwnProperty.call(obj, prop);
}

// Walks the type graph and explicitly propagates the type of aval to all
// downstream abstract values. This is stored in the recv property.
function propagateAndStoreExplicitly(aval) {
  // Gets all abstract values that this abstract value (aval) propagates to
  function getForwardsRecursively(av, seen) {
    if (!seen) seen = [];
    if (av.forward) for (var i = 0; i < av.forward.length; i++) {
      var fwd = av.forward[i].propagatesTo();
      if (!fwd) continue;
      var dest;
      if (fwd instanceof infer.AVal) dest = fwd;
      else if (fwd.target instanceof infer.AVal) {
        // TODO(sqs): Handle propagations to properties of types that may or may
        // not exist concretely. (fwd.pathExt is the key path.) See `function
        // expore` in tern/lib/infer.js for how this is done in tern's inference engine.
      } else continue;
      if (dest && seen.indexOf(dest) == -1) {
        seen.push(dest);
        getForwardsRecursively(dest, seen);
      }
    }
    return seen;
  }

  var allFwds = getForwardsRecursively(aval);
  for (var i = 0; i < allFwds.length; i++) {
    var fwd = allFwds[i];
    if (fwd == aval) continue;
    if (!fwd.recv) fwd.recv = [];
    fwd.recv.push(aval);
  }
}

function describe(path, aval, state) {
  var o = {id: path, key: symbol_id.parse(path)};

  if (aval.recv) o.recv = aval.recv.map(function(av) { return require('./annotate').getTarget(av); }).sort();

  var type = aval.getType(true);
  if (type) {
    o.type = infer.toString(type, 1);
  }

  if (aval.doc) o.doc = aval.doc;

  var node = aval.originNode;
  if (node) {
    o.file = astNodeFilename(node);

    // Track whether idents are of the definition of a symbol or just a reference.
    node._isDef = true;

    o.ident = state.getSpan(node);

    try {
      var defn = defnode.findDefinitionNode(node.sourceFile.ast, node.start, node.end);
      if (defn) o.defn = state.getSpan({start: Math.min(node.start, defn.start), end: Math.max(node.end, defn.end)});
    } catch (e) {}
  }

  o.exported = path.indexOf('@') == -1;
  o.data = aval.metaData;

  return o;
}

function store2(path, info, state) {
  if (!info.relevant) return;
  var out = {path: path, id: symbol_id.parse(path)};
  var name = typeName(info.type);
  if (name != info.type.path && name != '?') {
    out.type = name;
  } else if (info.type.proto && info.type.proto != state.cx.protos.Object) {
    var protoName = typeName(info.type.proto);
    if (protoName != '?') out.proto = protoName;
  }
  if (info.file) out.file = info.file;
  if (info.type) {
    out.typeDef = {file: info.type.origin};
    var bodySpan = state.getSpan(info.type);
    if (bodySpan) out.typeDef.bodySpan = bodySpan;
    if (info.type.originNode && info.type.originNode._identNodes) {
      out.typeDef.identSpans = info.type.originNode._identNodes.map(function(n) { return state.getSpan({originNode: n}); });
    }
  }
  if (info.object) {
    var objFile;
    if (info.object.originNode) {
      objFile = info.object.originNode.sourceFile.name;
    }
    out.objectDef = {file: objFile};
    var identSpan = state.getSpan(info.object);
    if (identSpan) out.objectDef.identSpan = identSpan;
    if (info.object.originNode && info.object.originNode._bodyNode) out.objectDef.bodySpan = state.getSpan({originNode: info.object.originNode._bodyNode});
  }
  if (info.doc) out.doc = info.doc;
  if (info.data) out.data = info.data;
  out.exported = path.slice(1).indexOf('!') == -1;
  state.output.push(out);
}

function astNodeFilename(node) {
  // Sometimes a node doesn't have a sourceFile property but one of its
  // child nodes does. In that case, get sourceFile from the child node.
  // TODO(sqs): why does this occur?
  if (!node.sourceFile) {
    for (var prop in node) if (node.hasOwnProperty(prop) && prop != 'type' && prop != 'start' && prop != 'end' && prop != 'scope') {
      var filename;
      if (node[prop] != node) filename = astNodeFilename(node[prop]);
      if (filename) return filename;
    }
  } else return node.sourceFile.name;
}

function runPass(functions) {
  if (functions) for (var i = 0; i < functions.length; ++i)
    functions[i].apply(null, Array.prototype.slice.call(arguments, 1));
}
